### PR TITLE
Refactor hash.h for CUDA

### PR DESCRIPTION
**Pull request by timgrant**
## Description

I have added ``HOSTDEVICE`` decorators to hash.h to pave the way for using the OSL noise generation functions on CUDA devices.

Other minor changes include guarding some SIMD-enabled color conversion functions in color.h, and making use of the ``__funnelshift`` intrinsic in fmath.h.

## Tests

No explicit tests added. Indirect testing through exercising the noise functions in OSL.

## Checklist:

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [x] If this is more extensive than a small change to existing code, I
  have previously submitted a Contributor License Agreement
  ([individual](../src/doc/CLA-INDIVIDUAL), and if there is any way my
  employers might think my programming belongs to them, then also
  [corporate](../src/doc/CLA-CORPORATE)).
- [x] I have updated the documentation, if applicable.
- [ ] I have ensured that the change is tested somewhere in the testsuite
  (adding new test cases if necessary).
- [x] My code follows the prevailing code style of this project.

### CODE REVIEW

File: src/include/OpenImageIO/hash.h [hunk 1]
```diff
@@ -564,17 +564,17 @@ class OIIO_API SHA1 {
     };
 
     /// Get the digest and store it in Hash h.
-    void gethash (Hash &h);
+    OIIO_HOSTDEVICE void gethash (Hash &h);
```
Comment (lgritz) [inline src/include/OpenImageIO/hash.h]: The SHA1 ones certainly aren't used for noise.

**Comment by timgrant** [comment]
I took a fairly mechanical "in for a penny" approach, and it's no trouble scaling it back.

**Comment by lgritz** [comment]
Let's just say, ok for the inline functions, but pointless to decorate the declarations that are implemented elsewhere and not compiled for Cuda currently.

**Comment by timgrant** [comment]
Sounds like a good rule of thumb. I'll update.

**Comment by lgritz** [comment]
LGTM!

**Pull request merged**